src/emulate.c

```diff
@@ -1125,146 +1126,201 @@ static bool rv_emulate(struct riscv_t *rv, struct rv_insn_t *ir)
     }
 
     /* step over instruction */
-    rv->PC += rv->insn_len;
+    rv->PC += ir->insn_len;
     return true;
 }
 
-void rv_step(struct riscv_t *rv, int32_t cycles)
+static bool insn_is_branch(uint8_t opcode)
 {
-    assert(rv);
-    uint32_t insn;
-    struct rv_insn_t ir;
-    const uint64_t cycles_target = rv->csr_cycle + cycles;
-
-    while (rv->csr_cycle < cycles_target && !rv->halt) {
-        /* enforce zero register */
-        rv->X[rv_reg_zero] = 0;
-
-        /* fetch the next instruction */
-        insn = rv->io.mem_ifetch(rv, rv->PC);
-        memset(&ir, 0, sizeof(struct rv_insn_t));
-
-        /* decode the instruction */
-        if (!rv_decode(&ir, insn, &rv->insn_len)) {
-            rv_except_illegal_insn(rv, insn);
-            break;
-        }
-
-        /* execute the instruciton */
-        if (!rv_emulate(rv, &ir))
-            break;
-
-        /* increment the cycles csr */
-        rv->csr_cycle++;
+    switch (opcode) {
+    case rv_insn_jal:
+    case rv_insn_jalr:
+    case rv_insn_beq:
+    case rv_insn_bne:
+    case rv_insn_blt:
+    case rv_insn_bge:
+    case rv_insn_bltu:
+    case rv_insn_bgeu:
+    case rv_insn_ecall:
+    case rv_insn_ebreak:
+    case rv_insn_mret:
+#if RV32_HAS(EXT_C)
+    case rv_insn_cj:
+    case rv_insn_cjr:
+    case rv_insn_cjal:
+    case rv_insn_cjalr:
+    case rv_insn_cbeqz:
+    case rv_insn_cbnez:
+    case rv_insn_cebreak:
+#endif
+#if RV32_HAS(Zifencei)
+    case rv_insn_fencei:
+#endif
+        return true;
     }
+    return false;
 }
 
-riscv_user_t rv_userdata(struct riscv_t *rv)
+/* hash function is used when mapping address into the block map */
+static uint32_t hash(size_t k)
 {
-    assert(rv);
-    return rv->userdata;
+    k ^= k << 21;
+    k ^= k >> 17;
+#if (SIZE_MAX > 0xFFFFFFFF)
+    k ^= k >> 35;
+    k ^= k >> 51;
+#endif
+    return k;
 }
 
-bool rv_set_pc(struct riscv_t *rv, riscv_word_t pc)
+/* allocate a basic block */
+static struct block *block_alloc(const uint8_t bits)
 {
-    assert(rv);
-#if RV32_HAS(EXT_C)
-    if (pc & 1)
-#else
-    if (pc & 3)
-#endif
-        return false;
-
-    rv->PC = pc;
-    return true;
+    struct block *block = malloc(sizeof(struct block));
+    block->insn_capacity = 1 << bits;
+    block->insn_number = 0;
+    block->predict = NULL;
+    block->ir = malloc(block->insn_capacity * sizeof(struct rv_insn_t));
+    return block;
 }
 
-riscv_word_t rv_get_pc(struct riscv_t *rv)
+/* insert a block into block map */
+static void block_insert(struct block_map *map, struct block *block)
```

Comment:
Add modifier `const` to `block` since it would be unchanged in this function.

Reply:
I will rewrite as following because the member in `block` would be unchanged
```diff
- static void block_insert(struct block_map *map, struct block *block)
+ static void block_insert(struct block_map *map, const struct block *block)
```
Ditto, other functions like `block_find` and `block_emulate` as well.